### PR TITLE
fix: handle optional map types with `GenericWriter[any]` (issue 415)

### PR DESCRIPTION
**Pull request by MaudGautier**
## Related issue

See issue https://github.com/parquet-go/parquet-go/issues/415

## Problem

When we use GenericWrite[any] with a schema where a map field is marked with the `optional` tag, but the actual data uses `*map[K]V`, the writer panics with:
```
panic: reflect: call of reflect.Value.Len on ptr to non-array Value [recovered]
	panic: reflect: call of reflect.Value.Len on ptr to non-array Value
```

<details> 
<summary>Expand to see the full stack trace:</summary>

```
panic: reflect: call of reflect.Value.Len on ptr to non-array Value [recovered]
	panic: reflect: call of reflect.Value.Len on ptr to non-array Value

goroutine 27 [running]:
testing.tRunner.func1.2({0x100abda60, 0x100c7ed10})
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/testing/testing.go:1734 +0x1ac
testing.tRunner.func1()
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/testing/testing.go:1737 +0x334
panic({0x100abda60?, 0x100c7ed10?})
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/runtime/panic.go:792 +0x124
reflect.Value.lenNonSlice({0x100a8f8a0?, 0x140000b2168?, 0x10830e918?})
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/reflect/value.go:1780 +0x220
reflect.Value.Len(...)
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/reflect/value.go:1761
github.com/parquet-go/parquet-go.writeValueFuncOfMap.func1({0x1400016ddc0, 0x2, 0x2}, {0x0, 0x0, 0x1}, {0x100a8f8a0?, 0x140000b2168?, 0x1?})
	/Users/maud/Documents/projects/ext/parquet-go/column_buffer_reflect.go:528 +0x1e4
github.com/parquet-go/parquet-go.writeValueFuncOfOptional.func1({0x1400016ddc0, 0x2, 0x2}, {0x38?, 0x1a?, 0xb?}, {0x100a8f8a0?, 0x140000b2168?, 0x14000159188?})
	/Users/maud/Documents/projects/ext/parquet-go/column_buffer_reflect.go:341 +0xc8
github.com/parquet-go/parquet-go.writeValueFuncOfGroup.func1({0x1400016ddc0, 0x2, 0x2}, {0xc0?, 0xdd?, 0x16?}, {0x100b24240, 0x140000b2168, 0x19})
	/Users/maud/Documents/projects/ext/parquet-go/column_buffer_reflect.go:635 +0x998
github.com/parquet-go/parquet-go.NewGenericWriter[...].func2({0x4c, 0xfe, 0x1a}, {{0x1400017ca30?, 0x100bcd780?, 0x0?}})
	/Users/maud/Documents/projects/ext/parquet-go/writer.go:149 +0xcc
github.com/parquet-go/parquet-go.NewGenericWriter[...].func4({0x1400017ca30, 0x1?, 0x1})
	/Users/maud/Documents/projects/ext/parquet-go/writer.go:212 +0x108
github.com/parquet-go/parquet-go.(*GenericWriter[...]).Write.func1(0x140002250b0?)
	/Users/maud/Documents/projects/ext/parquet-go/writer.go:242 +0x68
github.com/parquet-go/parquet-go.(*writerRowGroup).writeRows(0x140000bd7c0, 0x1, 0x14000113e90)
	/Users/maud/Documents/projects/ext/parquet-go/writer.go:918 +0x88
github.com/parquet-go/parquet-go.(*GenericWriter[...]).Write(0x100ca5e40, {0x1400017ca30, 0x1, 0x1})
	/Users/maud/Documents/projects/ext/parquet-go/writer.go:241 +0xa0
github.com/parquet-go/parquet-go.TestGenericWriterAnyPointerVsOptionalTag.func1(0x1400020a700)
	/Users/maud/Documents/projects/ext/parquet-go/column_buffer_reflect_test.go:1849 +0xc8
testing.tRunner(0x1400020a700, 0x14000224f90)
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/testing/testing.go:1792 +0xe4
created by testing.(*T).Run in goroutine 21
	/Users/maud/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.9.darwin-arm64/src/testing/testing.go:1851 +0x374

```

</details>

## Root cause

This occurred because the returned writeValueFunc of `writeValueFuncOfMap` does not handle the case where `mapValue` is a pointer to a map (so it calls `.Len()` on a pointer which is forbidden. Note: this occurs only with non-nil data - if the data is nil, there is already a guard for it in https://github.com/parquet-go/parquet-go/blob/027e5bd5b3003ba7820eb559bc6d136122c5a821/column_buffer_reflect.go#L498-L502).

## Solution proposed in this PR

In `writeValueFuncOfMap`, if `mapValue` is a pointer, dereference it.

## Test cases

Test cases for optional strings and optional int (and any other optional scalar) already passed before the fix in that PR. They're here to show the pattern of having an optional XX with a `optional` tag in the schema and a pointer in the data is supported in those cases.

The case `map_nil` where we have an optional map and a nil value also passed, because it was handled by this code so that we don't fall in the panic for `.Len()`: https://github.com/parquet-go/parquet-go/blob/027e5bd5b3003ba7820eb559bc6d136122c5a821/column_buffer_reflect.go#L498-L502

The cases that were failing were `map_value` and `map_non_nil_pointer_to_nil_map`, i.e. when the data we pass is non null.

## Note

What I'm fixing in this PR is a regression that occurred somewhere in between v0.25.1 and the current code version (having an optional map with some non nil key-values worked before).

### CODE REVIEW

File: column_buffer_reflect.go [hunk 1]
```diff
@@ -495,6 +495,10 @@ func writeValueFuncOfMap(columnIndex int16, node Node) (int16, writeValueFunc) {
 	zeroKeyValue := reflect.Zero(keyValueElem)
 
 	return nextColumnIndex, func(columns []ColumnBuffer, levels columnLevels, mapValue reflect.Value) {
+		if mapValue.Kind() == reflect.Pointer {
+			mapValue = mapValue.Elem()
```
Comment (achille-roussel) [inline column_buffer_reflect.go:504]: I believe this would panic if the pointer was `nil`, could we check and handling the same way than nil maps are handled below?

Comment (MaudGautier) [reply column_buffer_reflect.go:504]: Hey, thanks a lot for reviewing!

Sure, I added an early return in this commit: https://github.com/parquet-go/parquet-go/pull/416/changes/fed39e9275acefa5b53254607105275dca320cc4

FYI I also already had a test case for the `nil` pointer case [here](https://github.com/MaudGautier/parquet-go/blob/655b4aa6f2f821ed037a9a62beb17163979bae9b/column_buffer_reflect_test.go#L1816-L1824)

----
Side-note FYI: I checked the docstrings, and it seems Elem() does not panic on a nil pointer but instead returns a zero value (which is already handled below with the `mapValue.IsValid()`):
```
// Elem returns the value that the interface v contains
// or that the pointer v points to.
// It panics if v's Kind is not [Interface] or [Pointer].
// It returns the zero Value if v is nil.
func (v Value) Elem() Value {
```
I don't mind at all keeping the early return though, it's really up to you!

Comment (achille-roussel) [reply column_buffer_reflect.go:504]: Oh nice, I was mistaken!

Thanks for making the update.

**Review by achille-roussel — APPROVED** [review]
